File: frontend/src/services/communityService.ts
import { api } from '@/lib/api'
import type { CommunityCategory, CommunityPost, CommunityPostComment, PaginatedResponse } from '@/types'

export const communityService = {
  categories: () => api.get<{ data: CommunityCategory[] }>('/community/categories'),

  list: (params?: { category?: string; search?: string; municipality_id?: number; page?: number; archived?: boolean }) =>
    api.get<PaginatedResponse<CommunityPost>>('/community/posts', {
      params: {
        ...params,
        archived: params?.archived ? 1 : undefined,
      },
    }),

  feed: (params?: { category?: string; search?: string; page?: number }) =>
    api.get<PaginatedResponse<CommunityPost>>('/community/feed', { params }),

  get: (id: number) => api.get<{ data: CommunityPost }>(`/community/posts/${id}`),

  like: (id: number) => api.post<{ message: string; data: CommunityPost }>(`/community/posts/${id}/like`),

  share: (id: number) => api.post<{ message: string; data: CommunityPost }>(`/community/posts/${id}/share`),

  comments: (id: number) => api.get<{ data: CommunityPostComment[] }>(`/community/posts/${id}/comments`),

  addComment: (id: number, body: string, parentId?: number, image?: File) => {
    if (image) {
      const formData = new FormData()
      formData.append('body', body)
      if (parentId != null) {
        formData.append('parent_id', String(parentId))
      }
      formData.append('image', image)
      return api.post<{ message: string; data: CommunityPostComment }>(`/community/posts/${id}/comments`, formData, {
        headers: { 'Content-Type': 'multipart/form-data' },
      })
    }

    return api.post<{ message: string; data: CommunityPostComment }>(`/community/posts/${id}/comments`, {
      body,
      parent_id: parentId,
    })
  },

  create: (payload: {
    title: string
    content: string
    category: string
    is_published?: boolean
    municipality_id?: number
    image?: File
  }, rolePrefix: 'mao' | 'ppo' | 'admin') => {
    if (payload.image) {
      const formData = new FormData()
      formData.append('title', payload.title)
      formData.append('content', payload.content)
      formData.append('category', payload.category)
      formData.append('is_published', payload.is_published !== false ? '1' : '0')
      if (payload.municipality_id != null) {
        formData.append('municipality_id', String(payload.municipality_id))
      }
      formData.append('image', payload.image)
      return api.post<{ message: string; data: CommunityPost }>(`/${rolePrefix}/community/posts`, formData, {
        headers: { 'Content-Type': 'multipart/form-data' },
      })
    }

    const { image: _, ...jsonPayload } = payload
    return api.post<{ message: string; data: CommunityPost }>(`/${rolePrefix}/community/posts`, jsonPayload)
  },

  archive: (id: number, rolePrefix: 'mao' | 'ppo' | 'admin') =>
    api.post<{ message: string }>(`/${rolePrefix}/community/posts/${id}/archive`),
  restore: (id: number, rolePrefix: 'mao' | 'ppo' | 'admin') =>
    api.post<{ message: string }>(`/${rolePrefix}/community/posts/${id}/restore`),
}

File: mobile/lib/services/community_service.dart
import 'package:dio/dio.dart';

import '../core/api/api_client.dart';
import '../models/community_post.dart';
import '../models/post_comment.dart';

class CommunityService {
  CommunityService(this._api);

  final ApiClient _api;

  Future<List<CommunityPost>> listPosts({String? category, String? search}) {
    return _api.handle(
      _api.get('/community/posts', queryParameters: {
        if (category != null) 'category': category,
        if (search != null && search.isNotEmpty) 'search': search,
      }),
      (json) {
        final data = (json as Map<String, dynamic>)['data'] as List<dynamic>;
        return data.map((e) => CommunityPost.fromJson(e as Map<String, dynamic>)).toList();
      },
    );
  }

  Future<List<CommunityPost>> feed({String? category, String? search}) {
    return _api.handle(
      _api.get('/community/feed', queryParameters: {
        if (category != null) 'category': category,
        if (search != null && search.isNotEmpty) 'search': search,
      }),
      (json) {
        final data = (json as Map<String, dynamic>)['data'] as List<dynamic>;
        return data.map((e) => CommunityPost.fromJson(e as Map<String, dynamic>)).toList();
      },
    );
  }

  Future<CommunityPost> like(int id) {
    return _api.handle(
      _api.post('/community/posts/$id/like'),
      (json) => CommunityPost.fromJson((json as Map<String, dynamic>)['data'] as Map<String, dynamic>),
    );
  }

  Future<CommunityPost> share(int id, {String? caption}) {
    return _api.handle(
      _api.post('/community/posts/$id/share', data: {
        if (caption != null && caption.isNotEmpty) 'caption': caption,
      }),
      (json) => CommunityPost.fromJson((json as Map<String, dynamic>)['data'] as Map<String, dynamic>),
    );
  }

  Future<CommunityPost> getPost(int id, {int? shareId}) {
    return _api.handle(
      _api.get('/community/posts/$id', queryParameters: {
        if (shareId != null) 'share_id': shareId,
      }),
      (json) => CommunityPost.fromJson((json as Map<String, dynamic>)['data'] as Map<String, dynamic>),
    );
  }

  Future<List<PostComment>> comments(int postId) {
    return _api.handle(
      _api.get('/community/posts/$postId/comments'),
      (json) {
        final data = (json as Map<String, dynamic>)['data'] as List<dynamic>;
        return data.map((e) => PostComment.fromJson(e as Map<String, dynamic>)).toList();
      },
    );
  }

  Future<PostComment> addComment(
    int postId, {
    String? body,
    int? parentId,
    String? imagePath,
  }) async {
    if (imagePath != null && imagePath.isNotEmpty) {
      final form = FormData.fromMap({
        if (body != null && body.trim().isNotEmpty) 'body': body.trim(),
        if (parentId != null) 'parent_id': parentId,
        'image': await MultipartFile.fromFile(imagePath),
      });

      return _api.handle(
        _api.postMultipart('/community/posts/$postId/comments', form),
        (json) => PostComment.fromJson((json as Map<String, dynamic>)['data'] as Map<String, dynamic>),
      );
    }

    return _api.handle(
      _api.post('/community/posts/$postId/comments', data: {
        'body': body?.trim() ?? '',
        if (parentId != null) 'parent_id': parentId,
      }),
      (json) => PostComment.fromJson((json as Map<String, dynamic>)['data'] as Map<String, dynamic>),
    );
  }
}
